commons/zenoh-shm/src/api/provider/shared_memory_provider.rs

```diff
@@ -0,0 +1,910 @@
+    _phantom3: PhantomData<DeallocatePolicy>,
+}
+impl<const N: usize, InnerPolicy, AltPolicy, DeallocatePolicy> AllocPolicy
+    for Deallocate<N, InnerPolicy, AltPolicy, DeallocatePolicy>
+where
+    InnerPolicy: AllocPolicy,
+    AltPolicy: AllocPolicy,
+    DeallocatePolicy: ForceDeallocPolicy,
+{
+    fn alloc<IDSource: ProtocolIDSource, Backend: SharedMemoryProviderBackend>(
+        layout: &MemoryLayout,
+        provider: &SharedMemoryProvider<IDSource, Backend>,
+    ) -> ChunkAllocResult {
+        let mut result = InnerPolicy::alloc(layout, provider);
+        for _ in 0..N {
+            match result {
+                Err(ZAllocError::NeedDefragment) | Err(ZAllocError::OutOfMemory) => {
+                    if !DeallocatePolicy::dealloc(provider) {
+                        return result;
+                    }
+                }
+                _ => {
+                    return result;
+                }
+            }
+            result = AltPolicy::alloc(layout, provider);
+        }
+        result
+    }
+}
+
+/// Blocking allocation policy.
+/// This policy will block until the allocation succeeds.
+/// Both sync and async modes available.
+#[zenoh_macros::unstable_doc]
+pub struct BlockOn<InnerPolicy = JustAlloc>
+where
+    InnerPolicy: AllocPolicy,
+{
+    _phantom: PhantomData<InnerPolicy>,
+}
+#[async_trait]
+impl<InnerPolicy> AsyncAllocPolicy for BlockOn<InnerPolicy>
+where
+    InnerPolicy: AllocPolicy,
+{
+    async fn alloc_async<
+        IDSource: ProtocolIDSource + Send + Sync,
+        Backend: SharedMemoryProviderBackend + Sync,
+    >(
+        layout: &MemoryLayout,
+        provider: &SharedMemoryProvider<IDSource, Backend>,
+    ) -> ChunkAllocResult {
+        loop {
+            match InnerPolicy::alloc(layout, provider) {
+                Err(ZAllocError::NeedDefragment) | Err(ZAllocError::OutOfMemory) => {
+                    // todo: implement provider's async signalling instead of this!
+                    tokio::time::sleep(Duration::from_millis(1)).await;
+                }
+                other_result => {
+                    return other_result;
+                }
+            }
+        }
+    }
+}
+impl<InnerPolicy> AllocPolicy for BlockOn<InnerPolicy>
+where
+    InnerPolicy: AllocPolicy,
+{
+    fn alloc<IDSource: ProtocolIDSource, Backend: SharedMemoryProviderBackend>(
+        layout: &MemoryLayout,
+        provider: &SharedMemoryProvider<IDSource, Backend>,
+    ) -> ChunkAllocResult {
+        loop {
+            match InnerPolicy::alloc(layout, provider) {
+                Err(ZAllocError::NeedDefragment) | Err(ZAllocError::OutOfMemory) => {
+                    // todo: implement provider's async signalling instead of this!
+                    std::thread::sleep(Duration::from_millis(1));
+                }
+                other_result => {
+                    return other_result;
+                }
+            }
+        }
+    }
+}
+
+// todo: allocator API
```

Comment:
Is this some dead code no longer valid or is it kept on how to potentially implement it?

Reply:
This code is intentionally kept as a model for future allocator implementation. I also use it to validate SHM API for suitability for `allocator_api2::alloc::Allocator` trait requirements and types (e.g. `Layout`).